dipy/align/tests/test_mattes.py

```diff
@@ -0,0 +1,776 @@
+        metric = MattesBase(32)
+        metric.setup(static, moving, smask, mmask)
+
+        # 1. Update the joint distribution
+        sp_to_moving = np.linalg.inv(moving_aff)
+        samples_moving_grid = (sp_to_moving.dot(samples.T).T)[...,:dim]
+        intensities_moving, inside = interp_method(moving.astype(np.float32),
+                                                   samples_moving_grid)
+        intensities_moving = np.array(intensities_moving, dtype=np.float64)
+        metric.update_pdfs_sparse(intensities_static, intensities_moving)
+        J0 = np.copy(metric.joint)
+
+        # 2. Update the joint distribution gradient (the derivative of each
+        # histogram cell w.r.t. the transform parameters). This requires
+        # among other things, the spatial gradient of the moving image.
+        theta = transform.get_identity_parameters().copy()
+        grid_to_space = np.eye(dim + 1)
+        spacing = np.ones(dim, dtype=np.float64)
+        shape = np.array(static.shape, dtype=np.int32)
+        mgrad, inside = vf.gradient(moving.astype(np.float32), moving_aff,
+                                    spacing, shape, grid_to_space)
+        metric.update_gradient_dense(theta, transform, static.astype(np.float64),
+                                     moving.astype(np.float64), grid_to_space,
+                                     mgrad, smask, mmask)
+
+        # 3. Update the metric (in this case, the Mutual Information) and its
+        # gradient, which is computed from the joint density and its gradient
+        metric.update_mi_metric(update_gradient=True)
+
+        # Now we can extract the value and gradient of the metric
+        # This is the gradient according to the implementation under test
+        val0 = metric.metric_val
+        actual = np.copy(metric.metric_grad)
+
+        # Compute the gradient using finite-diferences
+        n = transform.get_number_of_parameters()
+        expected = np.empty_like(actual)
+        for i in range(n):
+            dtheta = theta.copy()
+            dtheta[i] += h
+
+            T = transform.param_to_matrix(dtheta)
+            shape = np.array(static.shape, dtype=np.int32)
+            sp_to_moving = np.linalg.inv(moving_aff).dot(T)
+            samples_moving_grid = (sp_to_moving.dot(samples.T).T)[...,:dim]
+            intensities_moving, inside = interp_method(moving.astype(np.float32),
+                                                       samples_moving_grid)
+            intensities_moving = np.array(intensities_moving, dtype=np.float64)
+            metric.update_pdfs_sparse(intensities_static, intensities_moving)
+            metric.update_mi_metric(update_gradient=False)
+            val1 = metric.metric_val
+            expected[i] = (val1 - val0) / h
+
+        dp = expected.dot(actual)
+        enorm = np.linalg.norm(expected)
+        anorm = np.linalg.norm(actual)
+        nprod = dp / (enorm * anorm)
+        assert_equal(nprod > 0.99, True)
+
```

Comment:
Thanks for checking so carefully Elef! I merged a couple of dense/sparse tests (test_mattes_densities() now checks both). I tried to do the same to test the gradients, but the resulting function would be very confusing. The general testing strategy (compare analytic vs. numeric derivatives, as Demian suggested) is the same (that's why they look so similar), but the exact procedure to interpolate and compute derivatives is different for dense and sparse, see for example:
https://github.com/omarocegueda/dipy/blob/imaffine_pr2/dipy/align/tests/test_mattes.py#L535
and 
https://github.com/omarocegueda/dipy/blob/imaffine_pr2/dipy/align/tests/test_mattes.py#L629
the first computes the dense gradient of the moving image and the second computes the gradient only at the sampling points. If I try to merge the two tests, the resulting function might be very confusing. What do you think? is this acceptable? or do you think it is worth trying to simplify more?